File: server/Services/PokemonService.cs
using PokeIpsum;
using PokeIpsum.Server.Models;
using Microsoft.Extensions.Caching.Memory;
using System.Globalization;

public class PokemonService
{
    private readonly HttpClient _http;
    private readonly IMemoryCache _cache;

    private readonly string _pokeapiURL = "https://pokeapi.co/api/v2/pokemon/";

    public PokemonService(HttpClient http, IMemoryCache cache)
    {
        _http = http;
        _cache = cache;
    }

    public async Task<PokemonDTO> ObterPokemonPorId(int id)

    {
        return await Commons.TratarRespostaAPI<PokemonDTO>(await _http.GetAsync($"{_pokeapiURL}{id}"));
    }

    public async Task<PokemonDTO> ObterPokemonPorNome(string nome)
    {
        return await Commons.TratarRespostaAPI<PokemonDTO>(await _http.GetAsync($"{_pokeapiURL}{nome}"));
    }

    public async Task<List<PokemonDTO>> ObterTodosOsPokemons()
    {
        var chaveCache = $"todos-os-pokemons_{DateTime.Now.ToString("MMMM", new CultureInfo("pt-BR"))}-{DateTime.Now.Year}";

        if (!_cache.TryGetValue(chaveCache, out List<PokemonDTO>? pokemons))
        {

            var listaPokemonsDTO = await Commons.TratarRespostaAPI<ListaPokemonDTO>(await _http.GetAsync($"{_pokeapiURL}?limit=99999"));
            pokemons = listaPokemonsDTO.Results;

            _cache.Set(chaveCache, pokemons);
        }
        return pokemons ?? [];
    }
}

File: server_nodejs/services/pokemonService.js
import axios from 'axios';
import NodeCache from 'node-cache';
import dotenv from 'dotenv';
import { extrairNomes } from '../utils/commons.js';

dotenv.config();

const cache = new NodeCache();
const pokeapiURL = `${process.env.POKEAPI_URL}pokemon/`;

async function obterPokemonPorId(id) {
    const url = `${pokeapiURL}${id}`;
    const response = await axios.get(url);
    if (response.data) {
        return { nome: response.data.name };
    }
    throw new Error('No data returned');
}

async function obterPokemonPorNome(nome) {
    const url = `${pokeapiURL}${nome}`;
    const response = await axios.get(url);
    if (response.data) {
        return { nome: response.data.name };
    }
    throw new Error('No data returned');
}

async function obterTodosOsPokemons() {
    const chaveCache = new Date().toLocaleString('pt-BR', { month: 'long', year: 'numeric' });
    let pokemons = cache.get(chaveCache);

    if (!pokemons) {
        const url = `${pokeapiURL}?limit=99999`;
        const response = await axios.get(url);
        if (response.data && response.data.results) {
            pokemons = extrairNomes(response.data.results);
            cache.set(chaveCache, pokemons, 2592000);
        } else {
            throw new Error('No data returned');
        }
    }
    return pokemons || [];
}

export { obterPokemonPorId, obterPokemonPorNome, obterTodosOsPokemons };